### PR TITLE
Issue #8 - (phase 1) script based war deployment and testing

### PR DESCRIPTION
**Pull request by joakime**
This is phase 1 for the automatic deployment of webapps and integration testing against the jetty9 base image.

More webapps and cleaner integeration testing will be subsequent phases.

### CODE REVIEW

**Comment by joakime** [comment]
Consider this script and demo webapp just a P.O.C

File: tests/run.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+docker tag "jetty9:latest" "gcr.io/${gcloud_projectid}/jetty9:testing"
+gcloud docker push "gcr.io/${gcloud_projectid}/jetty9:testing"
+
+# Deploy and Execute the individual test-war-* tests
+
+for module in test-war-*
+do
+  echo "module: $module"
+  appengine_deploy "${module}"
+done
```
Comment (aslo) [inline tests/run.sh]: It might be a good idea to remove the deployed version after the tests have run.
```
gcloud app versions delete "${test_runid}"
```

File: tests/test-war-hello/src/test/java/com/google/cloud/runtimes/jetty/tests/DeploymentITCase.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            IO.copy(reader, writer);
+            return writer.toString();
+        }
+    }
+    
+    @Test
+    public void testGet() throws IOException
+    {
+        HttpURLConnection http = (HttpURLConnection) serverURI.resolve("/hello/").toURL().openConnection();
```
Comment (aslo) [inline tests/test-war-hello/src/test/java/com/google/cloud/runtimes/jetty/tests/DeploymentITCase.java]: Do you envision that all test webapps will implement all the same routes?

File: tests/run.sh (hunk 1 above)
Comment (joakime) [inline tests/run.sh]: Noted.

Do you know how to create an machine/instance (or whatever the correct name is) with a specific environment configurations? (like say 2GB or 5GB for the webapp?)

File: tests/test-war-hello/src/test/java/com/google/cloud/runtimes/jetty/tests/DeploymentITCase.java (hunk 2 above)
Comment (joakime) [inline tests/test-war-hello/src/test/java/com/google/cloud/runtimes/jetty/tests/DeploymentITCase.java]: This is just a P.o.C.

I plan on using maven for it all (no shell scripts), and the integration-testing phase across each test-webapp, and will share a common testing core.

File: tests/run.sh (hunk 1 above)
Comment (aslo) [inline tests/run.sh]: It's possible to tweak the VMs via some settings in app.yaml. [See docs](https://cloud.google.com/appengine/docs/flexible/java/configuring-your-app-with-app-yaml). Is that what you mean?

Comment (joakime) [inline tests/run.sh]: I'll try the `app.yaml` approach ...

``` yaml
resources:
  cpu: .5
  memory_gb: 1.3
  disk_size_gb: 10
```

This would be good for testing the `GAE_MEMORY_MB` issue as well.

Comment (aslo) [inline tests/run.sh]: True, though if you just wanted to test that the jetty runtime obeys the `GAE_MEMORY_MB` env variable, you could just override that variable with an arbitrary value in a custom dockerfile, and then verify that the jvm was started with max heap set to 80% of that value. 

I don't think we need to worry about testing that the `GAE_MEMORY_MB` value is set properly for various VM configurations; I would consider that to be on the app engine platform's side of the contract. @meltsufin Do you have any thoughts on this?

Comment (meltsufin) [inline tests/run.sh]: I agree that we don't need to verify the value of GAE_MEMORY_MB because that may change. We just want to make sure that it is set, and that we are assigning 80% of it to JVM heap.

**Comment by meltsufin** [comment]
I think this is on the right track continuing what @aslo started. I like the actual tests being in Java and Maven rather than shell scripts, as long as it can be done with a reasonable amount of effort. Ultimately, we may be able to get by without using gcloud CLI at all and use APIs directly for interacting with cloud services.

File: tests/pom.xml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            <groupId>org.apache.maven.surefire</groupId>
+            <artifactId>surefire-junit47</artifactId>
+            <version>2.19.1</version>
+          </dependency>
+        </dependencies>
+      </plugin>
+      <plugin>
+        <groupId>com.google.cloud.tools</groupId>
+        <artifactId>appengine-maven-plugin</artifactId>
+        <version>0.1.1-beta</version>
```
Comment (meltsufin) [inline tests/pom.xml]: The latest version is 0.1.2

Comment (joakime) [inline tests/pom.xml]: That will be a separate PR for issue #22

File: tests/run.sh (hunk 1 above)
Comment (joakime) [inline tests/run.sh]: Point is, we need to test the environment behavior as well.   If we deploy in flex or not, the behavior is different, and is dependent on the memory_gb setting, so having at least the 4 different memory configurations is a good thing to test and ensure that we dont have a regression (which is the root goal of this issue and initial PR)

**Comment by meltsufin** [comment]
Is Alex's comment on `gcloud app versions delete "${test_runid}"` the only thing we're blocking on before committing this?

**Comment by joakime** [comment]
The versions delete will come in with subsequent commit / pull request

**Comment by meltsufin** [comment]
If the app doesn't successfully come up, will the Maven build fail?

**Comment by meltsufin** [comment]
Closing in favor of #45.
